xain_fl/coordinator/coordinator.py

```diff
@@ -236,6 +238,17 @@ def select_participant_ids_and_init_round(self) -> None:
         selected_ids = self.controller.select_ids(self.participants.ids())
         self.round = Round(selected_ids)
 
+    def select_outstanding(self) -> List[str]:
+        """Selects participant ids."""
+
+        selected = set(self.round.participant_ids)
+        num_outstanding = self.minimum_participants_in_round - len(selected)
+        pool = set(self.participants.ids()) - selected
```

Comment:
Are we sure this is always strictly positive? Otherwise the `frac` calculation below might raise a `DivisionByZero` or be negative.

Reply:
good question. here's an attempted mini-proof:

This function currently only gets called when there aren't enough participants selected _and_ there are enough connected to select from:
```python
len(self.round.participant_ids) < self.minimum_participants_in_round and
self.participants.len() == self.minimum_connected_participants             # (1)
```
and it should always be the case that
```python
self.minimum_participants_in_round <= self.minimum_connected_participants  # (2)
```
Now `num_outstanding` is necessarily positive due to `(1)` (assuming `minimum_participants_in_round` is, which should be!). Also `(1)` and `(2)` together imply
```python
len(self.round.participant_ids) < self.participants.len()
```
hence `pool` is also positive. Therefore `frac` is too.

Reply:
It may be worth assert a bunch of `assert` that checks our assumptions:

```python
assert(len(self.round.participant_ids) < self.minimum_participants_in_round)
# I think you meant >= here, instead of ==, right?
assert(self.participants.len() >= self.minimum_connected_participants   
# This one is so obvious I'm not sure this assertion is necessay
assert(self.minimum_participants_in_round <= self.minimum_connected_participants)
```

This would:

- catch a bug early if one of these invariants is violated
- help people reading/changing this function to understand under which condition it can be called

One additional note about (1):
```
len(self.round.participant_ids) < self.minimum_participants_in_round and
self.participants.len() == self.minimum_connected_participants
```

We must make sure that clients that sent their EndTraining message and that disconnected are kept in `self.round.participant_ids` but removed from `self.participants`. Can you confirm this is the case?

Reply:
yes I agree it would indeed be helpful to have some pre-conditions in there.

> `# I think you meant >= here, instead of ==, right?`

the naming of `minimum_connected_participants` is a little confusing at the moment as we shouldn't ever have more than this number connected. I believe the original idea was that there could potentially be more than this (and there is till a ticket for it), hence the naming.

> We must make sure that clients that sent their EndTraining message and that disconnected are kept in `self.round.participant_ids` but removed from `self.participants`.

Had thought about this as well, but decided against it for now after some deliberation. The thinking was, yes it would be better not to waste the work done by these guys, but I would consider this just as an optimisation over the more conservative approach of replacing them. Otherwise, it's an edge case that complicates the already quite complicated logic, and also the potential for `self.round.participant_ids` and `self.participants` to be out of sync might create issues down the line.